crates/net/discv5/src/lib.rs

```diff
@@ -0,0 +1,580 @@
+    /// An error from underlying [`discv5::Discv5`] node.
+    #[error("{0}")]
+    Discv5Error(discv5::Error),
+    /// An error from underlying [`discv5::Discv5`] node.
+    #[error("{0}")]
+    Discv5ErrorStr(&'static str),
+}
+
+/// Use API of [`discv5::Discv5`].
+pub trait HandleDiscv5 {
+    /// Exposes API of [`discv5::Discv5`].
+    fn with_discv5<F, R>(&self, f: F) -> R
+    where
+        F: FnOnce(&Discv5) -> R;
+
+    /// Returns the [`IpMode`] of the local node.
+    fn ip_mode(&self) -> IpMode;
+
+    /// Returns the key to use to identify the [`ForkId`] kv-pair on the [`Enr`](discv5::Enr).
+    fn fork_id_key(&self) -> &[u8];
+
+    /// Returns the [`ForkId`] of the given [`Enr`](discv5::Enr), if field is set.
+    fn get_fork_id<K: discv5::enr::EnrKey>(
+        &self,
+        enr: &discv5::enr::Enr<K>,
+    ) -> Result<ForkId, Error> {
+        let mut fork_id_bytes = enr.get_raw_rlp(self.fork_id_key()).ok_or(Error::ForkMissing)?;
+
+        Ok(ForkId::decode(&mut fork_id_bytes)?)
+    }
+
+    /// Tries to convert an [`Enr`](discv5::Enr) into the backwards compatible type [`NodeRecord`],
+    /// w.r.t. local [`IpMode`].
+    fn try_into_reachable(&self, enr: &discv5::Enr) -> Result<NodeRecord, Error> {
+        // todo: track unreachable with metrics
+        if enr.udp4_socket().is_none() && enr.udp6_socket().is_none() {
+            return Err(Error::UnreachableDiscovery)
+        }
+        let Some(udp_socket) = self.ip_mode().get_contactable_addr(enr) else {
+            return Err(Error::IpVersionMismatchDiscovery(self.ip_mode()))
+        };
+        // since we, on bootstrap, set tcp4 in local ENR for `IpMode::Dual`, we prefer tcp4 here
+        // too
+        let Some(tcp_port) = (match self.ip_mode() {
+            IpMode::Ip4 | IpMode::DualStack => enr.tcp4(),
+            IpMode::Ip6 => enr.tcp6(),
+        }) else {
+            return Err(Error::IpVersionMismatchRlpx(self.ip_mode()))
+        };
+
+        let id = uncompressed_id_from_enr_pk(enr);
+
+        Ok(NodeRecord { address: udp_socket.ip(), tcp_port, udp_port: udp_socket.port(), id })
+    }
+
+    /// Applies filtering rules on an ENR. Returns [`Ok`](FilterOutcome::Ok) if peer should be
+    /// passed up to app, and [`Ignore`](FilterOutcome::Ignore) if peer should instead be dropped.
+    fn filter_discovered_peer(&self, enr: &discv5::Enr) -> FilterOutcome;
+}
+
+/// Transparent wrapper around [`discv5::Discv5`].
+#[derive(Deref, DerefMut, Clone, Constructor)]
+pub struct Discv5 {
+    #[deref]
+    #[deref_mut]
+    discv5: Arc<discv5::Discv5>,
+    ip_mode: IpMode,
+    // Notify app of discovered nodes that don't have a TCP port set in their ENR. These nodes are
+    // filtered out by default. allow_no_tcp_discovered_nodes: bool,
+    fork_id_key: &'static [u8],
+    /// Optionally filter discovered peers before passing up to app.
+    discovered_peer_filter: MustNotIncludeChains,
+    #[doc(hidden)]
+    pub metrics: Metrics,
```

Comment:
does this need to be pub with hidden docs? would like to make private and remove `#[doc(hidden)]` if possible